### DISCUSSION TITLE
Path to a Dictionary Key

### DISCUSSION THREAD

**Discussion by ZiaMeza** [GodotNuts/GodotFirebase#418] [Q&A]
Hello, as usual, thanks for the work you've done to keep this updated. 

My question: is it possible to selectively update a single element in the dictionary stored on Firebase? 

In a game I'm working on, there is a dictionary I have that has hundreds of keys within it

word_experience: {가다: 6, 거기: 3, 기다리다: 1, 네: 6, 대: 6, 둘: 1, 먹다: 3, 멈추다: 3, 물: 28...

Updating on Firestore hasn't been a problem, as I basically do the equivalent of:

document.add_or_update_field(word_experience, {가다: 6, 거기: 3, 기다리다: 1, 네: 6, 대: 6, 둘: 1, 먹다: 3, 멈추다: 3, 물: 28...)

However, because there are multiple dictionaries like this and because saving happens so frequently, it stutters the game slightly when saving while sending the whole set of dictionaries. Therefore, instead of sending the whole dictionary, **I'd like to only send the section that changed without sending the whole dictionary.** For example,

document.add_or_update_field(word_experience, {가다: 50})

However, this outright replaces the dictionary that has been saved under the category, which is word_experience in this case. So I'm wondering if I can do something like this:

document.add_or_update_field(word_experience.가다, 50)

Where only the single key within word_experience is updated? I've rifled through the wiki and through Firebase documentation, but I'm unsure how to do something like this, or indeed whether it's possible. As always, thank you for your time.

**Comment by WolfgangSenff** [comment]
Hm. I'll test something. I think we can simplify your data structure and it'll make it much easier, but have to test it out.

**Reply by ZiaMeza** [reply]
That'd be swell! Should I provide a full example of the structure or any of my functions, or will this this independent testing on your end?

**Reply by WolfgangSenff** [reply]
Independent - just requires my putting in a huge thing. I assume the word_experience in your example is a constant that is a string though, right? Like the actual variable.

**Reply by ZiaMeza** [reply]
Gotcha. 

word_experience in this example is a dictionary, instantiated as:

`var word_experience = {}`

During saving, it is serialized as a dictionary as well, though I think it turns into a string during the... JSONification process. That's probably the right term for it.

```
func serialize_game_state() -> String:
	var game_state = {
		"word_experience": word_experience,
	}

	return game_state_json
```

**Reply by WolfgangSenff** [reply]
Yes, that is correct. (The word is correct, and it is both JSONification and stringified. :P) But in any case, yeah, I think what you need to do is move your dictionary to the top level instead. This means instead of having a variable in the document that has the dictionary in it, just have the document be the dictionary. So like this instead:
`
var document = await collection.add("WhateverDocumentName", { "word": frequency, "word2": frequency2 })
document.add_or_update_field(word, frequency + 1)
document = await collection.update(document)
` 

If you remove that root-level variable, it'll only "see" the changes that are made to the values in the document correctly. With the root-level variable, it is saying, "This whole thing has changed so I have to send/receive the entire thing," and with a really big dictionary, that's going to add up and start charging you too.

**Reply by WolfgangSenff** [reply]
So, that said, I can consider attempting to allow paths into there. I'll have to think about how to do it. Give me some time, and if you want, can you open a feature request for it?

**Reply by ZiaMeza** [reply]
Ohh I see. And I assume then I'd have to create a different document for each of the different variables (dictionaries), rather than storing all the dictionaries into one big mega document. So then each one can sort of automatically get the benefit of only "seeing" the changes, and thus would make the saves efficient enough to avoid the stuttering (probably).

If that sounds right, I'll use that as an implementation for now to make my saves more efficient, and I shall open the feature request as well since that kind of feature would still be useful in the future, especially with my dictionaries of dictionaries.

**Reply by WolfgangSenff** [reply]
Ah, yeah, if you have multiple sub-dictionaries like that, then it's harder. I have some time before my next interview, so I'll just try to implement it now and let you know how it goes.

**Reply by WolfgangSenff** [reply]
Okay, this is going to take quite a bit longer than I thought. You added the request which is awesome, but you probably should switch to what you were saying and use that in the meantime.

**Reply by ZiaMeza** [reply]
Sounds good, thanks for the prompt update! Take your time, as I believe I can reduce the scale issues quite significantly with your idea in the meantime, and so this thread has been a tremendous help already.
